transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,459 @@
+                            crypto,
+                        };
+                        let (id, connection) =
+                            match me.endpoint.connect(client_config, addr, "server_name") {
+                                Ok(c) => c,
+                                Err(err) => {
+                                    tracing::error!("dial failure: {}", err);
+                                    let _ = tx.send(Err(err.into()));
+                                    continue;
+                                }
+                            };
+                        let (channel, conn) = me.channel.create_connection(id);
+                        me.connections.insert(id, conn);
+                        let muxer = QuicMuxer::new(channel, connection);
+                        tx.send(Ok(muxer)).ok();
+                    }
+                    Some(ToEndpoint::ConnectionEvent {
+                        connection_id,
+                        event,
+                    }) => {
+                        let is_drained_event = event.is_drained();
+                        if is_drained_event {
+                            me.connections.remove(&connection_id);
+                        }
+                        if let Some(event) = me.endpoint.handle_event(connection_id, event) {
+                            if me.send_event(connection_id, event, cx) {
+                                tracing::info!("filled event slot");
+                                break;
+                            }
+                        }
+                    }
+                    Some(ToEndpoint::Transmit(transmit)) => {
+                        me.transmit(transmit);
+                    }
+                    None => {
+                        me.endpoint.reject_new_connections();
+                        return Poll::Ready(());
+                    }
+                }
+            }
+        }
+
+        while !me.outgoing.is_empty() {
+            me.outgoing.make_contiguous();
+            match me.socket.poll_send(cx, me.outgoing.as_slices().0) {
+                Poll::Ready(Ok(n)) => {
+                    me.outgoing.drain(..n);
+                }
+                Poll::Ready(Err(err)) => tracing::error!("send_to: {}", err),
+                Poll::Pending => break,
+            }
+        }
+
+        if me.event_slot.is_none() && me.incoming_slot.is_none() {
+            let mut metas = [RecvMeta::default(); BATCH_SIZE];
+            let mut iovs = MaybeUninit::<[IoSliceMut; BATCH_SIZE]>::uninit();
+            me.recv_buf
+                .chunks_mut(me.recv_buf.len() / BATCH_SIZE)
+                .enumerate()
+                .for_each(|(i, buf)| unsafe {
+                    iovs.as_mut_ptr()
+                        .cast::<IoSliceMut>()
+                        .add(i)
+                        .write(IoSliceMut::new(buf));
+                });
+            let mut iovs = unsafe { iovs.assume_init() };
```

Comment:
Why the unsafe code here?

Reply:
so `quinn` uses the same pattern. we allocate a vec at startup to store BATCH_SIZE udp packets. we then create a stack allocated array of pointers to pass to `recvmmsg`. we could heap allocate them in a vec instead so we don't have to declare uninitialized memory. however due to lifetime issues I don't think we can allocate the iovs at startup (at least not without more unsafe code).

Reply:
I suggest factoring this into a separate function, with a proof of safety that holds even if `IoSliceMut:new()` panics.

Reply:
We get UB if `chunks.len < BATCH_SIZE`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f53ebde1704cf8f6c70ade877ad17fc7

Reply:
yes, but it can't happen: `let recv_buf = vec![0; max_udp_payload_size * BATCH_SIZE].into_boxed_slice();`

Reply:
obviously if you assume uninitialized bytes are initialized that's UB

Reply:
Yes, you are right. There is a corner case when `max_udp_payload_size=0` but we will panic on `.chunks_mut(0)`.

Reply:
> with a proof of safety that holds even if `IoSliceMut:new()` panics.

@Demi-Marie `IoSliceMut:new()` does not panic. Also:

> Note that dropping a MaybeUninit<T> will never call T’s drop code. It is your responsibility to make sure T gets dropped if it got initialized.

Reply:
> I don't think we can allocate the iovs at startup (at least not without more unsafe code)

We allocate them but not initialize, because it's a self-reference issue.

Reply:
I think this is done? @Demi-Marie are you satisfied?

Reply:
Here is a PR https://github.com/dvc94ch/rust-libp2p/pull/17